pyra/hardware.py

```diff
@@ -0,0 +1,422 @@
+                dash_stats[f'gpu_{name}'].append(gpu_load)
+
+        amd_counter += 1
+
+
+def update_memory():
+    """
+    Update dashboard stats for system memory usage.
+
+    This will append a new value to the ``dash_stats['memory_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system memory percentage utilized.
+
+    Examples
+    --------
+    >>> update_memory()
+    """
+    memory_percent = psutil.virtual_memory().percent
+
+    if initialized:
+        dash_stats['memory_system'].append(memory_percent)
+
+    return memory_percent
+
+
+def update_network():
+    """
+    Update dashboard stats for system network usage.
+
+    This will append a new values to the ``dash_stats['network_received']`` and ``dash_stats['network_sent']``
+    lists.
+
+    Returns
+    -------
+    tuple
+        A tuple of the received and sent values as a difference since the last update.
+
+    Examples
+    --------
+    >>> update_network()
+    """
+    global initialized
+    global network_recv_last
+    global network_sent_last
+
+    net_recv_current = psutil.net_io_counters().bytes_recv / 1000000  # convert bytes to mb
+    net_sent_current = psutil.net_io_counters().bytes_sent / 1000000  # convert bytes to mb
+
+    net_recv_diff = net_recv_current - network_recv_last
+    net_sent_diff = net_sent_current - network_sent_last
+
+    network_recv_last = net_recv_current
+    network_sent_last = net_sent_current
+
+    if initialized:
+        dash_stats['network_received'].append(net_recv_diff)
+        dash_stats['network_sent'].append(net_sent_diff)
+
+    return net_recv_diff, net_sent_diff
+
+
+def update():
+    """
+    Update all dashboard stats.
+
+    This function updates the cpu and memory usage of this python process as well as subprocesses. Following that the
+    system functions are called to update system cpu, gpu, memory, and network usage. Finally the keys in the
+    ``dash_stats`` dictionary are cleaned up to only hold 120 values. This function is called once per second, so
+    therefore there are 2 minutes worth of values in the dictionary.
+
+    Examples
+    --------
+    >>> update()
+    """
+    global initialized
+
+    current_timestamp = helpers.timestamp()
+
+    if initialized:
+        dash_stats['timestamp'].append(helpers.timestamp())
+
+        dash_stats['relative_time'] = []
+        for x in dash_stats['timestamp']:
+            seconds_ago = current_timestamp - x
+            dash_stats['relative_time'].append(seconds_ago)
+
+        child_processes = proc.children(recursive=False)  # list all children processes
+
+        for child in child_processes:
+            if child not in processes:
+                processes.append(child)
+
+    for p in processes:
+        # set the name
+        proc_name = definitions.Names().name if p.pid == proc_id else p.name()
+
+        # cpu stats per process
+        proc_cpu_percent = None
```

Comment:
You could move this line to line 245. It would achieve the same affect with one less line in the code.

Reply:
I had this in the except statement before; however pycharm was indicating to me that there was a chance it could be undefined. Probably because my except statement is too specific. But pycharm cries if the except statement isn't specific, haha.

Figured moving it before would guarantee that it's defined.